[MO] - [testing doc] -> document about our tests architecture (#478)
strimzi/strimzi-kafka-bridge: development-docs/TESTING.md

# Testing Kafka Bridge

This document gives a detailed breakdown of the testing processes and testing options for `Kafka Bridge` within tests.

<!-- TOC depthFrom:2 -->

- [Pre-requisites](#pre-requisites)
- [Package Structure](#package-structure)
- [Approaches](#approaches)
- [Test Phases](#test-phases)
- [Available Test Groups](#available-test-groups)
- [Environment Variables](#environment-variables)
- [Running single test class](#running-single-test-class)

<!-- /TOC -->

## Pre-requisites

> To run any tests you need build whole project. You can achieve that with following command  `mvn clean install -DskipTests` 

Reviewer: ```suggestion
To run any tests you need to build whole project. You can achieve that with following command -  `mvn clean install -DskipTests` -
```

in root directory of the project.

The next requirement is to have [docker](https://docs.docker.com/get-docker/) installed, because we use test-containers. 
In case of interest you can check our implementation of Strimzi `Kafka` container [here](https://github.com/strimzi/strimzi-kafka-operator/tree/master/test-container).

## Package Structure

You can find tests inside `src.test.java` package. Moreover we have the auxiliary classes and the most notable one are:
 
- `src.test.java.io.strimzi.kafka.bridge.utils` = static methods
- `src.test.java.io.strimzi.kafka.bridge.clients` = clients for testing overall communication of the `Kafka Bridge`

> - `src.test.java.io.strimzi.kafka.bridge.facades` = encapsulation of the standalone `Kafka` and `AdminClient` instance

Reviewer 2: it's strange mixing folder path with package, I think we can avoid this and highlights only classes names `Utils`, `Clients` and `Facades` then using a list and not "=" :-)

Furthermore we divide our test suite on two separate parts:

1. `src.test.java.io.strimzi.kafka.bridge.http` = package, where you can find all test cases related to the `Http Kafka Bridge`

> 2. `src.test.java.io.strimzi.kafka.bridge.amqp` = package, where you can find all tests cases related to the `Amqp Kafka Bridge`

Reviewer 2: ditto as above

## Approaches

There are four approaches how you can verify that `Kafka Bridge` is without any bug. Before I list all approaches 
the following  terminology is needed.

#### In-memory 
`Kafka` = 
- for `Http Kafka Bridge` we use [Strimzi Kafka container]([here](https://github.com/strimzi/strimzi-kafka-operator/tree/master/test-container).)
- for `Amqp Kafka Bridge` we use [KafkaFacade](https://github.com/strimzi/strimzi-kafka-bridge/blob/master/src/test/java/io/strimzi/kafka/bridge/facades/KafkaFacade.java) 

`Kafka Bridge` =

>  -  we use for deployment `Http Kafka Bridge` and `Amqp Kafka Bridge` the [Vert.x](https://vertx.io/) 

Reviewer: ```suggestion
 - for deployment of the `Http Kafka Bridge` and `Amqp Kafka Bridge` we are using [Vert.x](https://vertx.io/) 
```
maybe?

#### Standalone 
`Kafka` 
 - we use latest version of the `Kafka` and you can start the `Kafka` instance by the `zookeeper-server-start.sh` and

> `kafka-server-start.sh`, which can be for instance download from [this](https://kafka.apache.org/downloads) site

Reviewer: ```suggestion
`kafka-server-start.sh`, which can be for instance downloaded from [this](https://kafka.apache.org/downloads) site
```

`Kafka Bridge` 
 - start `Kafka Bridge` instance using bash script located `target/kafka-bridge-{version}/kafka-bridge-{version}/bin/kafka_bridge_run.sh` 
with following command `kafka_bridge_run.sh --config-file ../config/application.properties`
 
----  
The approaches are as follows:

1. Using in-memory `Kafka` and in-memory `Kafka Bridge`
2. Using standalone `Kafka` and standalone `Kafka Bridge`
3. Using in-memory `Kafka` and standalone `Kafka Bridge`

> 4. Using standalone `Kafka` and in-memory `Kafka Bridge`

Reviewer 2: is it feasible to have a table 2 x 2 checking what is supported and what is not?

In our CI we are currently testing the first approach which is the most important. Also, in the future we are planning 
to add the second option. Rest of these approaches are not so essential. 

## Test Phases

In general, we use classic test phases: `setup`, `exercise`, `verify` and `teardown`. Every phase will be described for the
`src.test.java.io.strimzi.kafka.bridge.http` package. It is very similar with `src.test.java.io.strimzi.kafka.bridge.amqp` only
one difference is that `http` package using Strimzi `Kafka` container by `test-containers` and `amqp` package using KafkaFacade.

### Setup

In this phase we perform:

* Deploy the in-memory Strimzi `Kafka` container or standalone `Kafka`
* Deploy the AdminClient
* Deploy the in-memory `Kafka Bridge` or standalone `Kafka-Bridge`
* Deploy the WebClient

Everything is setup in the `HttpBridgeTestBase` class. In the static block we instantiate the Strimzi `Kafka` container,

> AdminClient to be shared across all test suites. In @Before we create our `Kafka` client, `Kafka Bridge` instance and lastly

Reviewer: ```suggestion
AdminClient to be shared across all test suites. In `@Before` we create our `Kafka` client, `Kafka Bridge` instance and lastly
```

        kafkaUri = kafkaContainer.getBootstrapServers();        <--- get Kafka bootstrap address

        adminClientFacade = AdminClientFacade.create(kafkaUri); <--- deploy in-memory AdminClient
    }
    ...
    ...
    @BeforeAll
    void setup() {
    
        basicKafkaClient = new BasicKafkaClient(kafkaUri);  <--- setup Kafka client
        bridgeConfig = BridgeConfig.fromMap(config);        <--- setup bridge config
        
        httpBridge = new HttpBridge(bridgeConfig, new MetricsReporter(jmxCollectorRegistry, meterRegistry)); <--- new instance of Kafka Bridge
        httpBridge.setHealthChecker(new HealthChecker());

        LOGGER.info("Deploying in-memory bridge");
        vertx.deployVerticle(httpBridge, context.succeeding(id -> context.completeNow())); <--- deploy in-memory Kafka Bridge
    
        client = WebClient.create(vertx, new WebClientOptions() <--- webclient for communication with REST API of Kafka Bridge
            .setDefaultHost(Urls.BRIDGE_HOST)
            .setDefaultPort(Urls.BRIDGE_PORT));
}
```

In case testing standalone Strimzi `Kafka` container it is excepted that you have already spin-up your Kafka cluster locally. 

> Same is applied on standalone `Kafka Bridge` instance. Note that you have to have set the environment variables `BRIDGE_EXTERNAL_ENV` and

Reviewer: ```suggestion
Same is applied on standalone `Kafka Bridge` instance. Note that you have to set the environment variables `BRIDGE_EXTERNAL_ENV` and
```

All available test groups are listed in [Constants](https://github.com/strimzi/strimzi-kafka-bridge/blob/master/src/test/java/io/strimzi/kafka/bridge/Constants.java) class.

## Environment variables

System tests can be configured by several environment variables, which are loaded before test execution.

| Name                                   | Description                                                                              | Default                                         |
| :------------------------------------: | :--------------------------------------------------------------------------------------: | :----------------------------------------------:|
| KAFKA_EXTERNAL_ENV                     |  Specify if tests should run against in-memory or external Kafka cluster                 | false                                           |
| BRIDGE_EXTERNAL_ENV                    | Specify if tests should run against in-memory or external Kafka Bridge                   | false                                           |

## Running single test class

Use the `verify` build goal and provide `-Dit.test=TestClassName[#testMethodName]` system property.

>     mvn verify -Dit.test=ConsumerTest#receiveSimpleMessage

Reviewer: ```suggestion
   ` mvn verify -Dit.test=ConsumerTest#receiveSimpleMessage`
```

 
----  
The approaches are as follows:

|  Approach	| Supported |
|---	|---	| 
|   1. Using in-memory `Kafka` and in-memory `Kafka Bridge` | Yes
|   2. Using standalone `Kafka` and standalone `Kafka Bridge` | Yes
|   3. Using in-memory `Kafka` and standalone `Kafka Bridge` | No
|   4. Using standalone `Kafka` and in-memory `Kafka Bridge` | No

In our CI we are currently testing the first approach which is the most important. Also, in the future we are planning 
to add the second option. Rest of these approaches are not so essential. 

## Test Phases

In general, we use classic test phases: `setup`, `exercise`, `verify` and `teardown`. Every phase will be described for the
`src.test.java.io.strimzi.kafka.bridge.http` package. It is very similar with `src.test.java.io.strimzi.kafka.bridge.amqp` only
one difference is that `http` package using Strimzi `Kafka` container by `test-containers` and `amqp` package using KafkaFacade.

### Setup

In this phase we perform:

* Deploy the in-memory Strimzi `Kafka` container or standalone `Kafka`
* Deploy the AdminClient
* Deploy the in-memory `Kafka Bridge` or standalone `Kafka-Bridge`
* Deploy the WebClient

> Everything is setup in the `HttpBridgeTestBase` class. In the static block we instantiate the Strimzi `Kafka` container,

Reviewer 3: Maybe we should rename `HttBridgeTestBase`to `HttpBridgeTestAbstract` since we use Abstrac naming in Strimzi as well. But it's just a note.

If you want specify the profile, use the `-P` flag - for example `-Phttpbridge`.

All available test groups are listed in [Constants](https://github.com/strimzi/strimzi-kafka-bridge/blob/master/src/test/java/io/strimzi/kafka/bridge/Constants.java) class.

## Environment variables

System tests can be configured by several environment variables, which are loaded before test execution.

| Name                                   | Description                                                                              | Default                                         |
| :------------------------------------: | :--------------------------------------------------------------------------------------: | :----------------------------------------------:|
| KAFKA_EXTERNAL_ENV                     |  Specify if tests should run against in-memory or external Kafka cluster                 | false                                           |
| BRIDGE_EXTERNAL_ENV                    | Specify if tests should run against in-memory or external Kafka Bridge                   | false                                           |

## Running single test class

> Use the `verify` build goal and provide `-Dit.test=TestClassName[#testMethodName]` system property.

Reviewer 3: Is this correct? I thought we use only surefire for Bridge tests.

Reviewer: when I tried that earlier, it didn't work, but I thought it's just some typo, can you take a look at it @see-quick?

Author: Yeah, `test` goal and `-Dtest` parameter should be used here thanks.
